# Assignment 4.8

**Question 1. Import libraries and ords_prods_merge dataframe.**

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create path to instacart basket analysis folder

path = r'/Users/mitchellkubik/Desktop/08-2023 Instacart Basket Analysis'

In [3]:
# Import 'ords_prods_merge' dataframe.

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', '4.7 Deriving New Variables', 'ords_prods_merged_new_variables'))

In [4]:
# Check the output

ords_prods_merge.shape

(32404859, 18)

In [5]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  mid range product   
1         Soda        77              7     9.0   both  mid range product   
2         Soda        77              7     9.0   both  mid range product   
3         Soda        77              7     9.0   both  mid range product   
4         Soda        77              7     9.0   both  mid range product   

      busiest_day    busiest_days busiest_period_of_day  
0  regularly busy  regularly busy        average orders  
1  regularly busy    slowest days        average orders  
2  regularly busy    slowest days           most orders  
3      least busy    slowest days        average orders  
4      least busy    slowest days           most orders

**Question 2. Find the aggregated mean of the "order_number" comlumn grouped by "department_id" for the entire dataframe**

In [6]:
# Apply the agg() function to obtain the mean values for the 'order_number' column based on 'department_id'

ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

**Question 3. Analyze the result and comment on differences between the subset and entire dataframe.**

- The first thing that I notice is that the subset containing only the 1st million rows of the dataframe on includes departments 4,7,13,14,16,17,19,&20 while the entire dataframe contains all departments 1-21. 
- The next thing I notice is that when comparing the departments that are included in both the dataframe and the subset, the mean "order_number" tends to go down slightly with the exception of "department_id" 17 which actually increases substantially by about 4.4 orders. 
- The minimum mean of the subset is 11.294069, the max of the subset is 19.463012.
- The minimum mean of the entire fataframe is 15.215751, the max is 22.902379.

**Question 4. Create a loyalty flag for existing customers using the transform() and loc() funtions.**

In [7]:
# Apply the transform function on the "order_number" column to generate the maximum orders for each user.

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
# Check the output.

ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  mid range product   
1          Soda        77              7     9.0   both  mid range product   
2          Soda        77              7     9.0   both  mid range product   
3          Soda        77              7     9.0   both  mid range product   
4          Soda        77              7     9.0   both  mid range product   
5          Soda        77              7     9.0   both  mid range product   
6          Soda        77              7     9.0   both  mid range product   
7          Soda        77              7     9.0   both  mid range product   
8          Soda        77              7     9.0   both  mid range product   
9          Soda        77              7     9.0   both  mid range product   
10         Soda        77              7     9.0   both  mid range product   
11         Soda        77              7     9.0   both  mid range product   
12         Soda        77              7     9.0   both  mid range product   
13         Soda        77              7     9.0   both  mid range product   
14         Soda        77              7     9.0   both  mid range product   

       busiest_day    busiest_days busiest_period_of_day  max_order  
0   regularly busy  regularly busy        average orders         10  
1   regularly busy    slowest days        average orders         10  
2   regularly busy    slowest days           most orders         10  
3       least busy    slowest days        average orders         10  
4  

In [9]:
ords_prods_merge['max_order'].value_counts()

99    1171333
8      811843
6      811396
9      810213
7      803979
       ...   
97      44949
98      44587
96      40453
2           6
1           5
Name: max_order, Length: 99, dtype: int64

In [10]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
..       ...      ...           ...                 ...                ...   
95   3226575      360             1                   5                 12   
96   1469869      377             3                   5                 17   
97   1927023      387             2                   4                 10   
98    858092      420             4                   1                 19   
99    119314      420             5                   3                 19   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
..                     ...         ...                ...        ...   
95                     NaN         196                  1          0   
96                     3.0         196                  9          0   
97                    22.0         196                  3          0   
98                    30.0         196                  2          0   
99                    16.0         196                  2          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  mid range product   
1          Soda        77              7     9.0   both  mid range product   
2          Soda        77              7     9.0   both  mid range product   
3          Soda        77              7     9.0   both  mid range product   
4          Soda        77              7     9.0   both  mid range product   
..          ...       ...            ...     ...    ...                ...   
95         Soda        77              7     9.0   both  mid range product   
96         Soda        77              7     9.0   both  mid range product   
97         Soda        77              7     9.0   both  mid range product   
98         Soda        77              7     9.0   both  mid range product   
99         Soda        77              7     9.0   both  mid range product   

       busiest_day    busiest_days busiest_period_of_day  max_order  
0   regularly busy  regularly busy        average orders         10  
1   regularly busy    slowest days        average orders         10  
2   regularly busy    slowest days           most orders         10  
3       least busy    slowest days        average orders         10  
4       least busy    slowest days           most orders         10  
..             ...             ...                   ...        ...  
95  regularly busy  regularly busy           most orders          3  
96  regularly busy  regularly busy        average orders          3  
97      least busy    slowest days           most orders          8  
98  regularly busy    busiest days        average orders         22  
99  regularly busy    slowest days        average orders         22  

[100 rows x 19 columns]

In [11]:
# Disregard max number of rows to display.

pd.options.display.max_rows = None

In [12]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [13]:
# Create a "loyalty flag" column to denote the amount of orders of each customer using the new "max_order" column.

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [14]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [15]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [16]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [17]:
# Index 'user_id', 'order_number', and 'loyalty_flag' to check the output of the new column.

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

**Question 5. Identify the spending habits of each of the three types of customers you just identified. The goal is to determine if there's a difference in the price of products bought by these three types of customers.**

In [18]:
ords_prods_merge.groupby('prices')

In [19]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

It looks like, for all intents and purposes, there is not a major difference in spending habits between new, regular, and loyal customers. I will note, however, that it looks like as customers order more products over time, their average amount spent on a given product decreases. 

**Question 6. Create a "spending flag" based on the avg price across each customers orders.**

In [20]:
# Start by creating a new 'avg_price' column based on user_id

ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [21]:
# Check the output

ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [22]:
# Check descriptive statistics on the prices column for each type of customer.

ords_prods_merge.groupby('loyalty_flag').agg({'avg_price' : ['mean','min','max']})

avg_price                        
                       mean       min           max
loyalty_flag                                       
Loyal customer    10.386336  2.387671   8794.882427
New customer      13.294670  1.000000  25005.425000
Regular customer  12.495717  1.308333   4506.269204

**I want to flag these max prices as suspiciously high for groceries ordered online. Owen, any insight here?**

In [23]:
# Now create a spending flag based on this new column

ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low Spender'

In [24]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High Spender'

In [25]:
# Check the output/frequencies of this new column in an efficient manner

ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low Spender     31770614
High Spender      634245
Name: spending_flag, dtype: int64

**Question 7. Create an order frequency flag that marks the regularity of a customers ordering according to the average "days_since_prior_order".**

In [26]:
# Create a new column to show a customers average length in between orders. Group by 'user_id'

ords_prods_merge['avg_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [27]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [28]:
# Create an order frequency flag that marks the regularity of a user's spending behavior using the loc() function 

ords_prods_merge.loc[ords_prods_merge['avg_days_between_orders'] > 20, 'frequency_flag'] = 'Non-Frequent Customer'

In [29]:
ords_prods_merge.loc[(ords_prods_merge['avg_days_between_orders'] > 10) & (ords_prods_merge['avg_days_between_orders'] <= 20), 'frequency_flag'] = 'Regular Customer'

In [30]:
ords_prods_merge.loc[ords_prods_merge['avg_days_between_orders'] <= 10, 'frequency_flag'] = "Frequent Customer"

In [31]:
# Check the frequency of each value

ords_prods_merge['frequency_flag'].value_counts(dropna=False)

Frequent Customer        21559853
Regular Customer          7208564
Non-Frequent Customer     3636437
NaN                             5
Name: frequency_flag, dtype: int64

In [32]:
# Check the output

ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

**Question 9. Export your dataframe as a pickle file and store it correctly in your "Prepared Data" folder.**

In [33]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4.8.pkl'))